providers/implementations/ciphers/cipher_ascon128.c

```diff
@@ -0,0 +1,710 @@
+            ok &= OSSL_PARAM_set_size_t(p, ASCON_AEAD_NONCE_LEN);
+        } else if (strcmp(p->key, OSSL_CIPHER_PARAM_AEAD_TAGLEN) == 0) {
+            ok &= OSSL_PARAM_set_size_t(p, ctx->tag_len);
+        } else if (strcmp(p->key, OSSL_CIPHER_PARAM_AEAD_TAG) == 0) {
+            /* Check that p->data_type matches "octet string" */
+            /* Check that p->data (the given buffer) is not NULL */
+            if (p->data == NULL || p->data_type != OSSL_PARAM_OCTET_STRING)
+            {
+                ok = 0;
+                break;
+            }
+
+            /* Check if the given buffer is big enough */
+            if (p->data_size < FIXED_TAG_LENGTH)
+            {
+                ok = 0;
+                break;
+            }
+
+            /* Check if ctx->is_tag_set is true */
+            if (!ctx->is_tag_set)
+            {
+                ERR_raise(ERR_LIB_PROV, PROV_R_TAG_NOT_SET);
+                ok = 0;
+                break;
+            }
+            /* Copy tag to destination */
+            memcpy(p->data, ctx->tag, FIXED_TAG_LENGTH);
+            p->return_size = FIXED_TAG_LENGTH;
+            ok &= 1;
+        } else if (strcmp(p->key, OSSL_CIPHER_PARAM_UPDATED_IV) == 0) {
+            /* Check that p->data_type matches "octet string" */
+            /* Check that p->data (the given buffer) is not NULL */
+            if (p->data == NULL || p->data_type != OSSL_PARAM_OCTET_STRING)
+            {
+                ok = 0;
+                break;
+            }
+
+            /* Check if the given buffer is big enough */
+            if (p->data_size < ASCON_AEAD_NONCE_LEN)
+            {
+                ok = 0;
+                break;
+            }
+
+            /* Check if ctx->iv_set is true */
+            if (!ctx->iv_set)
+            {
+                ok = 0;
+                break;
+            }
+            /* Copy IV to destination */
+            memcpy(p->data, ctx->iv, ASCON_AEAD_NONCE_LEN);
+            p->return_size = ASCON_AEAD_NONCE_LEN;
+            ok &= 1;
+        }
+    }
+
+    return ok;
+}
+
+/* Parameters that libcrypto can send to this implementation */
+static const OSSL_PARAM *ascon128_settable_ctx_params(ossl_unused void *cctx,
+                                                       ossl_unused void *provctx)
+{
+    static const OSSL_PARAM table[] = {
+        {OSSL_CIPHER_PARAM_AEAD_AAD, OSSL_PARAM_OCTET_STRING, NULL, 0, 0},
+        {OSSL_CIPHER_PARAM_AEAD_TAG, OSSL_PARAM_OCTET_STRING, NULL, 0, 0},
+        {OSSL_CIPHER_PARAM_AEAD_TAGLEN, OSSL_PARAM_UNSIGNED_INTEGER, NULL, sizeof(size_t), 0},
+        {NULL, 0, NULL, 0, 0},
+    };
+
+    return table;
+}
+
+static int ascon128_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    struct ascon_ctx_st *ctx = vctx;
+    const OSSL_PARAM *p;
+    int ok = 1;
+
+    if (ctx == NULL)
+    {
+        return 0;
+    }
+
+    for (p = params; p->key != NULL; p++) {
+        if (strcmp(p->key, OSSL_CIPHER_PARAM_AEAD_AAD) == 0) {
```

Comment:
Why is this param even needed?
Why not use the existing mechanism that allows update() to be called to add the AAD?